bindings/rust/evmc-declare/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+
+            result.into()
+        }
+    }
+}
+
+/// Takes an identifier and struct definition, builds an evmc_create_* function for FFI.
+fn build_create_fn(
+    name_lowercase: &String,
+    name_caps: &String,
+    type_name: &String,
+) -> proc_macro2::TokenStream {
+    let fn_name = format!("evmc_create_{}", name_lowercase);
+    let fn_ident = Ident::new(&fn_name, name_lowercase.span());
+    let type_ident = Ident::new(type_name, type_name.span());
+
+    // TODO: reduce code duplication here.
+    let capabilities_fn_string = format!("{}_get_capabilities", name_lowercase);
+    let capabilities_fn_ident = Ident::new(&capabilities_fn_string, name_lowercase.span());
+    let destroy_fn_string = format!("{}_destroy", name_lowercase);
+    let destroy_fn_ident = Ident::new(&destroy_fn_string, name_lowercase.span());
+    let static_name_string = format!("{}_NAME", name_caps);
+    let static_version_string = format!("{}_VERSION", name_caps);
+    let static_name_ident = Ident::new(&static_name_string, name_caps.span());
+    let static_version_ident = Ident::new(&static_version_string, name_caps.span());
+    let execute_fn_string = format!("{}_execute", name_lowercase);
+    let execute_fn_ident = Ident::new(&execute_fn_string, name_lowercase.span());
+
+    quote! {
+        #[no_mangle]
+        extern "C" fn #fn_ident() -> *const ::evmc_sys::evmc_instance {
+            let new_instance = ::evmc_sys::evmc_instance {
+                abi_version: ::evmc_sys::EVMC_ABI_VERSION as i32,
+                destroy: Some(#destroy_fn_ident),
+                execute: Some(#execute_fn_ident),
+                get_capabilities: Some(#capabilities_fn_ident),
+                set_option: None,
+                set_tracer: None,
+                name: ::std::ffi::CString::new(#static_name_ident).expect("Failed to build VM name string").into_raw() as *const i8,
+                version: ::std::ffi::CString::new(#static_version_ident).expect("Failed to build VM version string").into_raw() as *const i8,
+            };
+
+            __EvmcContainer::new::<#type_ident>(new_instance).into_ffi_pointer() as *const ::evmc_sys::evmc_instance
+        }
+    }
+}
+
+/// Builds a callback to dispose of the VM instance
+fn build_destroy_fn(name_lowercase: &String, type_name: &String) -> proc_macro2::TokenStream {
+    let fn_ident_string = format!("{}_destroy", name_lowercase);
+    let fn_ident = Ident::new(&fn_ident_string, name_lowercase.span());
+    let type_ident = Ident::new(type_name, type_name.span());
+
+    quote! {
+        extern "C" fn #fn_ident(instance: *mut ::evmc_sys::evmc_instance) {
+            Box::new(__EvmcContainer::from_ffi_pointer::<#type_ident>(instance));
+        }
+    }
+}
+
+/// Takes a capabilities flag and builds the evmc_get_capabilities callback.
+fn build_capabilities_fn(name_lowercase: &String, capabilities: u32) -> proc_macro2::TokenStream {
+    // Could avoid using a special name and just use get_capabilities.
```

Comment:
Yep, the special name is only required for `create` as that is the public symbol, everything else should be scoped for the file.

The only benefit we get with prefixes is that multiple VMs could be declared in the same file, but not sure we want that.

Reply:
Done.